app/packs/src/apps/mydb/collections/CollectionTreeStructure.js

```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import UserStore from '../../../stores/alt/stores/UserStore';
+
+const filterSharedWithMeCollection = (sharedCollections) => {
+  const { currentUser } = UserStore.getState();
+  const collections = [];
+
+  sharedCollections.forEach((collection) => {
+    let children = [];
+    let label = `by ${collection.shared_by.initials}`;
+    let user = {};
+    let uid = -1;
+    let sharedCollections = collection.collection_acls.filter(acl => (acl.user.id === currentUser.id ));
+    sharedCollections.forEach((acl) => {
+      children.push(acl);
+      user = acl.user;
+      uid = acl.id;
+    })
+
+    const sameSharedTo = collections.find(c => (c.label == label));
+    if (sameSharedTo) {
+      children.forEach(c => sameSharedTo.children.push(c))
+    } else {
+      let sharedCollection = {}
```

Comment:
here a new collection is created. I would extract that functionality to reduce complexity
and make it usable for the next function with also use this logic